typed-lambda-calculus/article.md

```diff
@@ -0,0 +1,717 @@
+
+A typed lambda calculus extends untyped lambda calculus with a type system.
+Note that unlike the untyped lambda calculus, there are multiple typed lambda calculi, differentiated by the features of the type system used.
+The exact features of the type system used can be chosen mostly freely, and we will discuss some common choices in this article.
+
+The untyped lambda calculus is hard to use as a logic system due to the paradoxes it creates.
+Adding Russell's type theory to lambda calculus makes it, under certain conditions, logically consistent, which in turn makes it possible to mathematically guarantee certain behaviors (or lack thereof).
+
+The most common argument for the strongly-typed languages is that if a program type checks, it will never segfault (or have any other type-related issues), but I think this argument sells typed languages short.
+A sufficiently strict and expressive type system allows for encoding much more than simply "not segfaulting".
+In some cases, the whole domain can be expressed in types.
+
+In this article, we're concerned with the basics, for the most part, so we won't get too much into the more expressive type systems like System FC or systems with dependent types.
+
+## Lambda calculus with atomic boolean values
+
+We will now introduce atomic values to the lambda calculus.
+While it is feasible to construct a value representation using only λ-abstractions (using Church encoding or otherwise) for almost any value we might be interested in, practically speaking, it's not very convenient.
+
+Thus, we introduce atomic values $true$ and $false$.
+
+Since we introduced atomic values, we also need to introduce an eliminator (i.e. a consumer) for these values.
+In this case let's consider an $\mathbf{if} \ldots \mathbf{then} \ldots \mathbf{else} \ldots$ operator.
+
+The semantics of this operator should be reasonably familiar to a practicing programmer, but let's informally write them them out anyway:
+
+- $\mathbf{if}\; true\; \mathbf{then}\;\alpha\;\mathbf{else} \;\beta$ should reduce to $\alpha,$
+- $\mathbf{if}\; false\; \mathbf{then}\;\alpha\;\mathbf{else} \;\beta$ should reduce to $\beta,$
+- $\mathbf{if}\; \gamma\; \mathbf{then}\;\alpha\;\mathbf{else} \;\beta$ where $\gamma$ is a redex, $\gamma$ should be reduced first.
+
+The introduction of the $\mathbf{if}$ operator creates a bit of a kink.
+What do we do if $\gamma$ is neither $true$, $false$, nor a redex?
+E.g. $\mathbf{if}\; (\lambda x. x)\; \mathbf{then}\;\ldots\;\mathbf{else} \;\ldots$
+
+Such terms are called "stuck", the computation, according to our rules, can not proceed.
+We want to avoid stuck terms!
+
+To do that, we'll have to place some restrictions on what $\gamma$ could be.
+
+### Typing relations, rules, and systems
+
+To express such restrictions, we introduce the so-called _typing relation_.
+A typing relation is a relation, in this case meaning a set of pairs, which assigns a _type_ to each and every expression in the language.
```

Comment:
When you say "each and every expression in the language", I suppose the idea is that only well-typed terms are "in the language", but then the statement becomes somewhat circular.

It would be more clear to consider all syntactically valid terms. Then you could say that the typing relation assign a type to _some_ expressions in the language, and those are considered well-typed. And those that do not have a type assigned to them by the typing relation are ill-typed.